### ISSUE TITLE
@auth with groupsField and groupClaim not working in V2 transformer

### ISSUE THREAD

**Issue by spc16670** [aws-amplify/amplify-cli#9788]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

14

### Amplify CLI Version

7.6.20

### What operating system are you using?

Windows

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

No manual changes made

### Amplify Categories

auth, api

### Amplify Commands

add

### Describe the bug

I upgraded from GraphQL transformer version: 1 to 2, updated my schema, deployed the changes and I see getting awful lot of 'Unauthorized' responses completely breaking my app in many places.

This is my model (one of them):

```js
type Client @model
  @auth(rules: [
    { allow: groups, groups: ["Root"] },
    { allow: groups, groups: ["Partner"], operations: [create] },
    { allow: groups, groupsField: "id", groupClaim: "client:owner" },
    { allow: groups, groupsField: "id", groupClaim: "client:custodian", operations: [read, update] },
    { allow: groups, groupsField: "id", groupClaim: "client:user", operations: [read] }
  ]) 
{
  id: ID! @primaryKey
  name: String!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime
}
```

I configure amplify to send ID token to the server

```js
Amplify.configure({
  API: {
    graphql_headers: async () => {
      const session = await Auth.currentSession();
      const idToken = session.getIdToken();
      return {
        Authorization: idToken.getJwtToken()
      };
    }
  },
});
```

My token has:

```js
client:custodian: "[\"GVP8919\"]"
```

There is the corresponding entry in the database with that ID.

And yet, the list operation does not return any results :frowning:
```js
{"data":{"listClients":{"__typename":"ModelClientConnection","items":[],"nextToken":null}}}
```

I am using Cognito user pools for authentication.

### Expected behavior

@auth with groupsField and groupClaim works the way it used to

### Reproduction steps

1. Update to V2
2. Create a model as described below
3. Login as user with Cognito Id token with custom claim as shown above
4. Attempt a list on the model

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line


```

</details>

### Log output

<details>

```
# Put your logs below this line


```

</details>

**Comment by yuth** [comment]
@spc16670 Its hard to understand the problem without the records and schemas.  Could you share GraphQL schemas with v1 and v2 which can be used to reproduce this issue. Please ensure there isn't any confidential info in the schema. 

Also include a few mutation which can be used to create records in V1 and query to retrieve those in V2.

**Comment by spc16670** [comment]
My v1 version of the schema is not very different:


```js
type Client @model 
  @auth(rules: [
    { allow: groups, groups: ["Root"] }
    { allow: groups, groups: ["Partner"], operations: [create] }
    { allow: groups, groupsField: "id", groupClaim: "client:owner" }
    { allow: groups, groupsField: "id", groupClaim: "client:custodian", operations: [read, update] }
    { allow: groups, groupsField: "id", groupClaim: "client:user", operations: [read] }
  ]) 
{
  id: ID!
  name: String!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime
}
```

I don't use any custom transformers - all stock amplify.

Is there any reason why the earlier schema and list operation done by a user with only ```client:custodian``` claim with a value of ```"[\"GVP8919\"]"``` should not work in v2?

**Comment by spc16670** [comment]
Why do you need mutation queries? They don't matter here, are you a member of amplify staff?

Here is the list query:

```js
query MyQuery {
  listClients {
    items {
      createdAt
      id
      name
      updatedAt
    }
  }
}
```

**Comment by hanna-becker** [comment]
We have the same issue after upgrading to the new transformer.

Simplified model:
```
type Project
@model
@auth(rules: [
  { allow: owner },
  { allow: groups, groupsField: "id", operations: [read, update] }
])
{
  id: ID!
  name: String!
}
```

In the PreTokenGenerator's alter-claims.js function, after fetching the projects the user should have access to, we're adding their ids as groups:
```
  event.response = {
      claimsOverrideDetails: {
        groupOverrideDetails: {
          groupsToOverride: projectIDs
        }
      }
  };
```

I can see that the group got added correctly to the client's JWT token, when decoded:
```
"cognito:groups": [
    "e3673234-bc4b-4700-892c-fa282fbc1919"
]
```

However, the user gets unauthorized errors for querying that project.

This used to work previously.

**Comment by alex-vladut** [comment]
I subscribe to this, I am also facing multiple issues in that area Not sure what Amplify team's plans are and if they are going to tackle it. I added a PR a while back to fix these issues but it seems they are not accepting any PRs at the moment due to some internal refactoring I understand https://github.com/aws-amplify/amplify-cli/pull/9535#issuecomment-1036700574

**Comment by endoodev1** [comment]
We are also having this problem as I have described in #9466 [here](https://github.com/aws-amplify/amplify-cli/pull/9466#issuecomment-1046260176) So I would also be more than happy to see a solution for this very soon! 
Moreover I think this problem needs to be stated somewhere in the [docs](https://docs.amplify.aws/cli/graphql/authorization-rules/#user-group-based-data-access) as a **"known limitation"** as long as this is not fixed. I think for most of the people running in this issue it is not very understandable. 
As some other guys (thanks @n0rb) found out, the bug comes up because the authentication vtl rules for an "id" field are created in [queries.ts](https://github.com/aws-amplify/amplify-cli/blob/1496724495010f2daec8b160e35e613ca34eaa5e/packages/amplify-graphql-auth-transformer/src/resolvers/query.ts) by _generateAuthOnModelQueryExpression_ method while the rules for the other "non id fields" were created by _generateAuthFilter_ method. 
Our model looks like this and has the same issue: 
```graphql
type Location @model 
@auth(rules: [ 
  { allow: groups, groups: ["Admin"], operations: [create, read, update, delete] } 
  { allow: groups, groupsField: "id", groupClaim: "la", operations: [read, update, delete] } 
]) {
  id: ID!
}
```
cc @danielleadams (I'm not sure you are the right contact person but you helped in #9535)

**Comment by hanna-becker** [comment]
Thanks for this comment, @endoodev1. With this information I was able to create a workaround for our use case until this issue is properly fixed. I added a new "group" property to the Project table that I duplicate the id value into. (This requires the Create query followed by and Update query, as the id is autogenerated, thus unknown at the time of creation.) At least this way we can get all our app's collab features working again and unblock development.
```
type Project
@model
@auth(rules: [
  { allow: owner },
  { allow: groups, groupsField: "group", operations: [read, update] }
])
{
  id: ID!
  group: String! # duplication of id for @auth group claim
  name: String!
}
```
Not sure group needs to necessarily be a required field, like I implemented it.

**Comment by danielleadams** [comment]
@spc16670 Thanks for reporting this (and thanks all for the chiming in) - I know you've provided a lot of information here, but can you provide specific reproduction steps to investigate this? I tried the following:

1. Create api with transformer v1, the group ID is set as the record ID. example schema:
```graphql
type Group @model @auth(rules: [
  { allow: owner, operations: [create, update, delete] }
  { allow: groups, groupsField: "id", operations: [read] }
]) {
  id: ID!
  name: String!
}
```
2. Create 2 users in cognito and add them to a group
3. Create records as user 1, set the record's group to the group's group name
4. Update API to use transformer v2 with the following schema:
```graphql
type Group @model @auth(rules: [
  { allow: owner }
  { allow: groups, groupsField: "id", operations: [read] }
]) {
  id: ID!
  name: String!
}
```
5. amplify push
6. Query list as user 1, returns list as expected with the previously created record
7. Query list as user 2, returns list with user 1's previously created record as expected

I've also tried variations of this, and I'm not able to reproduce. What am I missing here?

**Comment by hanna-becker** [comment]
We're having the issue with transformer v2. In v1 it worked fine.

**Comment by danielleadams** [comment]
@hanna-becker I am not able to reproduce the issue in v2, so I am asking for specific steps to duplicate the issue.

**Comment by hanna-becker** [comment]
Hi @danielleadams

Thanks for looking into this!

Re: step 2, I'm not sure I understood you correctly here, but do not add the users to a static group. Instead do this: When adding/updating auth, the cli asks about whether it should add triggers at some point. Add a PreTokenGeneration lambda trigger, and add the id of the shared resource (Group in your case, Project in my example above) in this block in the generated "alter-claims.js" file. 
```
      claimsOverrideDetails: {
        groupOverrideDetails: {
          groupsToOverride: [<insert shared resource id>]
        }
      }
```
Normally we would get the ids to add here from a DB query, but for the sake of reproducing this, hard-coding to the id of an existing record should be sufficient. 

In our case we recreated the app from scratch after the transformer upgrade, but this may be different for others, so I cannot say anything about whether or not creating the api with v1 and then updating to v2 will reproduce the issue. Maybe others can chime in here.

Otherwise the steps look like you should be able to reproduce the issue to me.

**Comment by endoodev1** [comment]
Hey @danielleadams, thanks for taking up this issue. 
As @hanna-becker already explained, it is a problem with the dynamic groups authentication policies in v2 transformer. 
In my previous [post](https://github.com/aws-amplify/amplify-cli/issues/9788#issuecomment-1046272215) I also explained the exact position in code where this bug comes from. It has to do with the vtl rules which are created in the [queries.ts](https://github.com/aws-amplify/amplify-cli/blob/1496724495010f2daec8b160e35e613ca34eaa5e/packages/amplify-graphql-auth-transformer/src/resolvers/query.ts) and which do not receive the groups claim from the jwt token as a json array in _generateAuthOnModelQueryExpression_ method when it is an id field. For other fields the vtl auth rules are created by _generateAuthFilter_ method which is working correctly (receiving the group claim as a json array) also because of pr #9466. In the PR this issue is explained much better. Unfortunately I'm not 100% sure how vtl language is working so I'm not able to create a good working pr on my own. 
This is vtl code of id field of _query.getXXX.auth.1.req.vtl_
``` vtl
#if( !$util.isNull($ctx.args.id) )
    #set( $idClaim = $util.defaultIfNull($ctx.identity.claims.get("GroupClaim-name"), "___xamznone____") )
    #if( $util.isString($ctx.args.id) )
      #set( $idCondition = ($idClaim == $ctx.args.id) )
    #else
      #set( $idCondition = ($idClaim == $util.defaultIfNull($ctx.args.id.get("eq"), "___xamznone____")) )
    #end
    #if( $idCondition )
      #set( $isAuthorized = true )
      $util.qr($ctx.stash.put("authFilter", null))
    #end
```
this is vtl code for non ID field: 
```vtl
#set( $role0 = $util.defaultIfNull($ctx.identity.claims.get("GroupClaim-name"), []) )
#if( $util.isString($role0) )
   #if( $util.isList($util.parseJson($role0)) )
      #set( $role0 = $util.parseJson($role0) )
   #else
        #set( $role0 = [$role0] )
   #end
#end 
#if( !$role0.isEmpty() )
      $util.qr($authFilter.add({ "tenant_id": { "in": $role0 } }))
#end
#if( !$authFilter.isEmpty() )
      $util.qr($ctx.stash.put("authFilter", { "or": $authFilter }))
#end
```

**Comment by endoodev1** [comment]
@Ilya93 this seems to be another issue. This issue here is about the specific case that you have the `groupsField=id`
In your case it's `groupsField=editors` However it seems to be a strange behavior as the update mutation should be part of the "update" operation. Have you checked that the group is set correctly in jwt token. What is unclear to me is how the BroadcastLiveData is created as there is no rule allowing "create". 
I would suggest to open another issue. 
@danielleadams are there any news when this issue will be addressed / fixed?

**Comment by spc16670** [comment]
@danielleadams I don't think that you have managed to reproduce in your example earlier what I have reported. Modify your example to use "groupsField" and "groupClaim" and ensure your id token has a custom group array populated as I have shown in my report.

It is concerning this has still not been marked as a bug and there is no acknowledgement from the staff. 

Until the issue is resolved the documentation here:

https://docs.amplify.aws/cli/graphql/authorization-rules/#how-it-works

should clarify that "groupsField" does not work "groupClaim".

Please refer to @endoodev1 said, he is reporting exactly the same issue stating where the problem is.

**Comment by danielleadams** [comment]
@hanna-becker @endoodev1 @spc16670 Thanks for the clarification. I think this is a duplicate of https://github.com/aws-amplify/amplify-cli/issues/9370. A contributor has been working on this, so it is in the process of getting fixed.

Please let me know if this does not seem correct or if you have any other thoughts. Thank you.

**Comment by mrsimply** [comment]
@danielleadams unfortunately with #9370 the current state is:
> [Straubulous](https://github.com/Straubulous) moved this from In progress to Stalled in [Bug bash](https://github.com/aws-amplify/amplify-cli/projects/7#card-77583773)

Is it possible you both somehow jointly have a look at this issue? 👀
Frankly in my opinion this really is a burning issue 🔥 for a lot of developers... Thank you in advance!

**Comment by IonasSoftwareLtd** [comment]
This is a burning issue - without this being fixed multi-tenancy in our app doesn't work. We are keeping everything on transformer V1 for now until the issue is resolved. This is a such a core feature of the framework I would imagine this kind of a bug should justify marking V2 as experimental - or the Amplify users will continue loosing confidence in the framework. 

Since Amplify is a part of AWS products family I would expect higher standards than from a regular open source project - and much quicker bug acknowledgment and resolution.
